File: app/src/main/java/com/gooner10/popularmoviesapp/data/MovieRepositoryImpl.java
package com.gooner10.popularmoviesapp.data;

import android.content.Context;
import android.support.annotation.NonNull;
import android.util.Log;

import java.util.List;

/**
 * Repository class to handle moviedata
 */

public class MovieRepositoryImpl implements IMovieRepository {
    public static final String TAG = MovieRepositoryImpl.class.getSimpleName();
    private static volatile MovieRepositoryImpl INSTANCE;
    private MovieDao movieDao;

    public MovieRepositoryImpl(MovieDao movieDao) {
        this.movieDao = movieDao;
    }

    public static MovieRepositoryImpl getMovieDatabaseInstance(@NonNull Context context) {
        if (INSTANCE == null) {
            synchronized (MovieRepositoryImpl.class) {
                if (INSTANCE == null) {
                    MovieDatabase movieDatabase = MovieDatabase.getInMemoryDatabase(context);
                    INSTANCE = new MovieRepositoryImpl(movieDatabase.movieDao());
                }
            }
        }
        return INSTANCE;
    }

    @Override
    public List<MovieItem> getMovie() {
        return movieDao.getAllFavoritMovie();
    }

    @Override
    public void insertOrUpdateMovieData(MovieItem movieData) {
        movieDao.insertAll(movieData);
    }

    @Override
    public void deleteAllMovieData(MovieItem movieData) {
        movieDao.deleteMovies(movieData);
    }

    @Override
    public boolean findMovieAlreadyIsFavorite(MovieItem movieData) {
        Log.d(TAG, "findMovieAlreadyIsFavorite: " + movieDao.movieExists(String.valueOf(movieData.getId())));
        return movieDao.movieExists(String.valueOf(movieData.getId()));
    }


}

File: core/data/src/main/java/com/gooner10/data/MovieRepositoryImpl.kt
package com.gooner10.data

import android.content.Context
import android.util.Log

/**
 * Repository class to handle moviedata
 */

class MovieRepositoryImpl(private val movieDao: MovieDao) : IMovieRepository {

    override val movie: List<com.gooner10.data.MovieItem>
        get() = movieDao.allFavoritMovie

    override fun insertOrUpdateMovieData(movieData: com.gooner10.data.MovieItem?) {
        movieDao.insertAll(movieData!!)
    }

    override fun deleteAllMovieData(movieData: com.gooner10.data.MovieItem?) {
        movieDao.deleteMovies(movieData!!)
    }

    override fun findMovieAlreadyIsFavorite(movieData: com.gooner10.data.MovieItem?): Boolean {
        Log.d(TAG, "findMovieAlreadyIsFavorite: " + movieDao.movieExists(movieData!!.id))
        return movieDao.movieExists(movieData.id)
    }

    companion object {
        val TAG = MovieRepositoryImpl::class.java.simpleName
        @Volatile
        private var INSTANCE: MovieRepositoryImpl? = null

        fun getMovieDatabaseInstance(context: Context): MovieRepositoryImpl? {
            if (INSTANCE == null) {
                synchronized(MovieRepositoryImpl::class.java) {
                    if (INSTANCE == null) {
                        val movieDatabase = MovieDatabase.getInMemoryDatabase(context)
                        INSTANCE = MovieRepositoryImpl(movieDatabase.movieDao())
                    }
                }
            }
            return INSTANCE
        }
    }


}
